daffodil-runtime2/src/main/resources/org/apache/daffodil/runtime2/c/libruntime/parsers.c

```diff
@@ -18,140 +18,484 @@
+
+    // Convert data endianness to host endianness
+    if (big_endian_data)
+    {
+        buffer.integer = be32toh(buffer.integer);
+    }
+    else
+    {
+        buffer.integer = le32toh(buffer.integer);
+    }
+
+    // Don't need to remove any padding bits because
+    // floats must be exactly 32 bits, otherwise SDE
+    assert(num_bits == sizeof(float) * BYTE_WIDTH);
+
+    // Return parsed number
+    *number = buffer.number;
+}
+
+// Helper method to read signed integers using fractional byte shifts
+// depending on data endianness (note not tested on big-endian
+// architecture; might work only on low-endian architecture)
+
+// When shifting signed integers, it is important that the type be
+// signed, otherwise the shift will not preserve the sign bit; this is
+// why we need a separate helper method for signed integers
+
+static void
+parse_endian_int64(bool big_endian_data, int64_t *number, size_t num_bits, PState *pstate)
+{
+    // Parse all signed integers in this helper function
+    union
+    {
+        uint8_t bytes[sizeof(int64_t)];
+        int64_t integer;
+    } buffer;
+
+    // Read data bits
+    buffer.integer = 0;
+    read_bits(buffer.bytes, num_bits, pstate);
+
+    // Shift data bits differently on endianness
+    if (big_endian_data)
+    {
+        // Convert data endianness to host endianness
+        buffer.integer = be64toh(buffer.integer);
+
+        // Need only num_bits so remove any padding bits
+        size_t shift = sizeof(int64_t) * 8 - num_bits;
+        buffer.integer >>= shift; // type must be signed
+    }
+    else
+    {
+        // Convert data endianness to host endianness
+        buffer.integer = le64toh(buffer.integer);
+
+        // Need only num_bits so remove any padding bits
+        size_t msb_ix = (num_bits - 1) / BYTE_WIDTH;
+        int8_t msb_byte = (int8_t)buffer.bytes[msb_ix];
+        size_t shift = (BYTE_WIDTH - num_bits % BYTE_WIDTH) % BYTE_WIDTH;
+        buffer.bytes[msb_ix] = msb_byte >> shift; // type must be signed
+
+        // Also, negative low-endian integers require setting
+        // following bytes to -1 to preserve the sign bit too
+        if (msb_byte < 0)
+        {
+            for (msb_ix++; msb_ix < sizeof(int64_t); msb_ix++)
+            {
+                buffer.bytes[msb_ix] = -1;
+            }
+        }
+    }
+
+    // Return parsed number
+    *number = buffer.integer;
+}
+
+// Helper method to read unsigned integers using fractional byte
+// shifts depending on data endianness (note not tested on big-endian
+// architecture; might work only on low-endian architecture)
+
+static void
+parse_endian_uint64(bool big_endian_data, uint64_t *number, size_t num_bits, PState *pstate)
+{
+    // Parse all unsigned integers in this helper function
+    union
+    {
+        uint8_t  bytes[sizeof(uint64_t)];
+        uint64_t integer;
+    } buffer;
+
+    // Read data bits
+    buffer.integer = 0;
+    read_bits(buffer.bytes, num_bits, pstate);
```

Comment:
Should these calls to `read_bits` check for an error and return? I guess technically it doesn't matter since this will just end up converting 0x0000... to a number, but that's just wasted cycles.

Reply:
I've made sure all calls to `read_bits` are followed by `if (pstate->error) return;` now (ditto for `write_bits` in unparsers.c).